### PR TITLE
[ONNX] Export index_fill and index_copy, fix caffe2 scatter

### PR DESCRIPTION
**Pull request by BowenBao**

### CODE REVIEW

**Review by houseroad — COMMENTED** [review]
rebase and resolve the conflicts?

**Comment by BowenBao** [comment]
> rebase and resolve the conflicts?

done, thanks

**Comment by BowenBao** [comment]
@bddppq please take a look, thanks.

**Review by bddppq — CHANGES_REQUESTED** [review]
Tests are failing:
https://circleci.com/api/v1.1/project/github/pytorch/pytorch/2503199/output/105/0?file=true
```
RuntimeError: ONNX export failed: Couldn't export operator aten::index_fill
```

**Comment by BowenBao** [comment]
@bddppq seems related to the change that removes CompleteTensorType. I'll update this PR. Thanks.

**Comment by BowenBao** [comment]
@dzhulgakov - please take a look at this one, thanks.

**Comment by houseroad** [comment]
fix conflicts?

**Comment by houseroad** [comment]
@BowenBao could you please fix the conflict?

File: caffe2/operators/utility_ops.h [hunk 1]
```diff
@@ -791,27 +791,33 @@ class ScatterOp : public Operator<CPUContext> {
...
-    for (int64_t batch = 0; batch < outer_dims_product; ++batch) {
-      int64_t i_max = i + N;
-      for (; i < i_max && i < indices.numel(); ++i) {
-        auto idx = idxs[i];
+    for (int64_t outer_batch = 0; outer_batch < outer_dims_product; ++outer_batch) {
+      for (int64_t i = 0; i < N; ++i) {
+        for (int64_t inner_batch = 0; inner_batch < idxs_block_size; ++inner_batch) {
+          auto idxs_elem_idx = outer_batch * idxs_batch_size + i * idxs_block_size + inner_batch;
+          auto src_elem_idx = outer_batch * src_batch_size + i * src_block_size + inner_batch;
+          auto dst_elem_idx = outer_batch * dst_batch_size + idxs[src_elem_idx] * dst_block_size + inner_batch;
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: idxs[src_elem_idx] ==> idxs[idxs_elem_idx]

btw, we should add some tests to cover such case, i.e., the dimensions are different.

File: caffe2/operators/utility_ops.h [hunk 2]
```diff
@@ -791,27 +791,33 @@ class ScatterOp : public Operator<CPUContext> {
...
 
-    check_indexarray_range<IndexType>(idxs, N, src_indexing_axis_dim);
+    check_indexarray_range<IndexType>(idxs, N, dst_indexing_axis_dim);
 
-    int64_t i = 0;
-    for (int64_t batch = 0; batch < outer_dims_product; ++batch) {
-      int64_t i_max = i + N;
-      for (; i < i_max && i < indices.numel(); ++i) {
-        auto idx = idxs[i];
+    for (int64_t outer_batch = 0; outer_batch < outer_dims_product; ++outer_batch) {
```
Comment (houseroad) [inline caffe2/operators/utility_ops.h]: Add some comments about how we compute the index for source and destination, otherwise it's really hard to follow.

**Review by houseroad — COMMENTED** [review]
Check my inline comments. Logic seems wrong. Let's fix it.

File: caffe2/operators/utility_ops.h (hunk 1 above)
Comment (BowenBao) [reply caffe2/operators/utility_ops.h]: Thanks for the catch, updated so that the code makes more sense. 
From execution-wise point of view it should be the same, since src and idxs have the same rank and shape. 

The test cases in test_pytorch_onnx_caffe2.py covers cases where input shape is different than update/indices shape.

File: caffe2/operators/utility_ops.h (hunk 2 above)
Comment (BowenBao) [reply caffe2/operators/utility_ops.h]: agreed... adding comments to help clarification

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -555,7 +578,7 @@ def forward(self, x, k):
     def test_layer_norm(self):
         model = torch.nn.LayerNorm([10, 10])
         x = torch.randn(20, 5, 10, 10)
-        self.run_test(model, x)
+        self.run_test(model, x, rtol=1e-05, atol=1e-06)
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: removed. also i think this is no longer needed or already fixed.

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -198,6 +198,9 @@ def _slice(g, input, axes, starts, ends):
         return input
     return g.op("Slice", input, axes_i=axes, starts_i=starts, ends_i=ends)
 
+def _constant_of_shape(g, shape, value, dtype):
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/symbolic_opset9.py]: unused, removed.

**Review by houseroad — APPROVED** [review]
Looks good now

**Comment by BowenBao** [comment]
Thanks @houseroad for reviewing!!
